objects/src/accounts/package/storage_entry/mod.rs

```diff
@@ -0,0 +1,477 @@
+                // Expect a Value variant: "slot" and "value" must be present
+                // "slots" and "values" must not be present.
+                if !slot_present {
+                    return Err(D::Error::custom("Missing 'slot' field for single-slot entry."));
+                }
+                if !value_present {
+                    return Err(D::Error::custom("Missing 'value' field for single-slot entry."));
+                }
+                Ok(StorageEntry::Value {
+                    name: raw.name,
+                    description: raw.description,
+                    slot: raw.slot.expect("was checked to be present"),
+                    value: raw.value.expect("was checked to be present"),
+                })
+            },
+            (true, false) => {
+                Err(D::Error::custom("`slots` is defined but no `values` field was found."))
+            },
+            (false, true) => {
+                // Expect a Map variant:
+                //   - `slot` must be present
+                //   - `values` must be present and convertible to map entries
+                //   - `slots` must not be present
+                //   - `value` must not be present
+                if !slot_present {
+                    return Err(D::Error::missing_field("slot"));
+                }
+                if raw.value.is_some() {
+                    return Err(D::Error::custom(
+                        "Fields 'value' and 'values' are mutually exclusive",
+                    ));
+                }
+
+                let values = raw.values.expect("was checked to be present");
+                let map_entries = values
+                    .into_map_entries()
+                    .ok_or_else(|| D::Error::custom("Invalid 'values' for map entry"))?;
+
+                Ok(StorageEntry::Map {
+                    name: raw.name,
+                    description: raw.description,
+                    slot: raw.slot.expect("was checked to be present"),
+                    values: map_entries,
+                })
+            },
+            (true, true) => {
+                // Expect a MultiSlot variant:
+                //   - `slots` must be present
+                //   - `values` must be present and represent words
+                //   - `slot` must not be present
+                //   - `value` must not be present
+                if slot_present {
+                    return Err(D::Error::custom(
+                        "Fields 'slot' and 'slots' are mutually exclusive.",
+                    ));
+                }
+                if value_present {
+                    return Err(D::Error::custom(
+                        "Fields 'value' and 'values' are mutually exclusive.",
+                    ));
+                }
+                let slots = raw.slots.ok_or_else(|| D::Error::missing_field("slots"))?;
+                let values = raw.values.ok_or_else(|| D::Error::missing_field("values"))?;
```

Comment:
If we change the match to be `match (raw.slots, raw.values)` we could deconstruct in the case to `(Some(slots), Some(values))` and remove these `ok_or_else`.